activity/ws/activity.go

```diff
@@ -48,25 +54,214 @@ func (a *Activity) Metadata() *activity.Metadata {
 // Eval implements api.Activity.Eval - Invokes a web socket operation
 func (a *Activity) Eval(ctx activity.Context) (done bool, err error) {
 	input := &Input{}
-	ctx.GetInputObject(input)
+	err = ctx.GetInputObject(input)
+	if err != nil {
+		return false, err
+	}
+	var isWSS bool
+	url := a.settings.URI
+	if url != "" {
+		isWSS = strings.HasPrefix(url, "wss")
+	}
+	parameters, err := GetParameter(ctx, input, ctx.Logger())
+	if err != nil {
+		ctx.Logger().Error(err)
+		return false, err
+	}
+	//populate custom headers
+	h := getHeaders(ctx, parameters)
+	//populate url with path and query params
+	builtURL := buildURI(url, parameters, ctx.Logger())
+	key := ctx.ActivityHost().Name() + "-" + ctx.Name() + "-" + builtURL + "-" + fmt.Sprintf("%v", h)
+	cachedConnection, ok := a.cachedClients.Load(key)
+	var connection *websocket.Conn
+	if !ok {
+		var dialer websocket.Dialer
+		if isWSS {
+			tlsconfig := &tls.Config{}
+			allowInsecure := a.settings.AllowInsecure
+			if allowInsecure {
+				tlsconfig.InsecureSkipVerify = true
+			} else {
+				var cacertObj map[string]interface{}
+				if a.settings.CaCert != "" {
+					err = json.Unmarshal([]byte(a.settings.CaCert), &cacertObj)
+					if err != nil { //file path configured
+						certPool, err := getCerts(a.settings.CaCert)
+						if err != nil {
+							fmt.Printf("Error while loading client trust store - %v", err)
+							return false, err
+						}
+						tlsconfig.RootCAs = certPool
+					} else { // file content configured
+						rootCAbytes, err := decodeCerts(a.settings.CaCert, ctx.Logger())
+						if err != nil {
+							ctx.Logger().Error(err)
+							return false, err
+						}
+						certPool := x509.NewCertPool()
+						certPool.AppendCertsFromPEM(rootCAbytes)
+						tlsconfig.RootCAs = certPool
+					}
+				}
+			}
+			dialer = websocket.Dialer{TLSClientConfig: tlsconfig}
+		} else {
+			dialer = *websocket.DefaultDialer
+		}
+		ctx.Logger().Info("Creating new connection")
+		conn, res, err := dialer.Dial(builtURL, h)
+		if err != nil {
+			if res != nil {
+				defer res.Body.Close()
+				body, err1 := ioutil.ReadAll(res.Body)
+				if err1 != nil{
+					ctx.Logger().Infof("res code is %v error while reading response payload is %s ", res.StatusCode,  err1)
```

Comment:
Here should be error logs right? ctx.Logger().Errorf

Reply:
done